Sources/Commands/Utilities/SymbolGraphExtract.swift

```diff
@@ -48,12 +57,115 @@ public struct SymbolGraphExtract {
+                verboseOutput: verboseOutput
+            )
+        }
+    }
+    
+    func extractSymbolGraphFromClangTarget(
+        target: ClangTarget,
+        buildDescription: ClangTargetBuildDescription,
+        outputRedirection: Process.OutputRedirection,
+        outputDirectory: AbsolutePath,
+        verboseOutput: Bool
+    ) throws {
+        // Collect the set of public headers by filtering all of the target's headers to those
+        // inside the target's include directory.
+        var publicHeaders = Set(target.headers.filter(target.includeDir.isAncestor))
+        
+        guard !publicHeaders.isEmpty else {
+            return
+        }
+        
+        let umbrellaHeader: AbsolutePath?
+        switch target.moduleMapType {
+        case .umbrellaHeader(let umbrellaHeaderFile):
+            // The target is configured with an umbrella header – store its path separately
+            // and remove it from the general list of public headers.
+            umbrellaHeader = umbrellaHeaderFile
+            publicHeaders.remove(umbrellaHeaderFile)
+        case .umbrellaDirectory(_):
+            // TODO: Support symbol graph extraction for umbrella directory module map type
+            fallthrough
+        case .custom(_):
+            // TODO: Support symbol graph extraction for custom module map file type
+            fallthrough
+        case .none:
+            umbrellaHeader = nil
+        }
+        
+        try fileSystem.createDirectory(outputDirectory, recursive: true)
+        
+        // Construct the command line arguments for extracting the symbol graph
+        var commandLine = [
+            clangCompiler.pathString, "-extract-api",
+            "--product-name=\(target.c99name)",
+            "-o", outputDirectory.appending(component: "\(target.name).symbols.json").pathString
+        ]
+        
+        if verboseOutput {
+            commandLine += ["-v"]
+        }
+        
+        switch outputFormat {
+        case .json(let pretty):
+            if pretty {
+                // FIXME: Specify pretty print behavior to `-extract-api` if it's supported.
+                // commandLine += ["-pretty-print"]
+            }
+        }
+        
+        // FIXME: Does it matter for `-extract-api` usage if the target is C++ or not?
+        commandLine += try buildDescription.basicArguments(isCXX: nil)
+        
+        // Pass the paths of all public headers to the extract-api command
+        commandLine += ["-x", "objective-c-header"]
+        
+        // If an umbrella header has been provided, pass it first
+        if let umbrellaHeader = umbrellaHeader {
+            commandLine += [umbrellaHeader.pathString]
+        }
+        // Then provide the remaining public headers in a deterministic order
+        commandLine += publicHeaders.lazy.map(\.pathString).sorted()
+        
+        // Run the extraction
+        let process = Process(
+            arguments: commandLine,
```

Comment:
Here's an example of what the command line arguments contain when running `swift package dump-symbol-graph` on my machine for the sample project I attached in the description:

```swift
[
  "/usr/bin/clang",
  "-extract-api",
  "--product-name=SampleObjCFramework",
  "-o",
  "/SampleObjCSwiftPackage/.build/x86_64-apple-macosx/symbolgraph/clang/SampleObjCFramework.symbols.json",
  "-fobjc-arc",
  "-target",
  "x86_64-apple-macosx10.13",
  "-g",
  "-O0",
  "-DSWIFT_PACKAGE=1",
  "-DDEBUG=1",
  "-fblocks",
  "-index-store-path",
  "/SampleObjCSwiftPackage/.build/x86_64-apple-macosx/debug/index/store",
  "-fmodules",
  "-fmodule-name=SampleObjCFramework",
  "-I",
  "/SampleObjCSwiftPackage/Sources/SampleObjCFramework/include",
  "-fmodules-cache-path=/SampleObjCSwiftPackage/.build/x86_64-apple-macosx/debug/ModuleCache",
  "-isysroot",
  "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk",
  "-F",
  "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks",
  "-fPIC",
  "-x",
  "objective-c-header",
  "/SampleObjCSwiftPackage/Sources/SampleObjCFramework/include/SampleObjCFramework.h",
  "/SampleObjCSwiftPackage/Sources/SampleObjCFramework/include/ImmutableRectangle.h",
  "/SampleObjCSwiftPackage/Sources/SampleObjCFramework/include/ShapeCollection.h",
  "/SampleObjCSwiftPackage/Sources/SampleObjCFramework/include/Triangle.h",
  "/SampleObjCSwiftPackage/Sources/SampleObjCFramework/include/TwoDimensionalShape.h"
]
```

Reply:
That looks about right. I don't think we need `-g`, `-O0` and `-fPIC` since we are only parsing ASTs but apart from that it seems correct.

Reply:
> I don't think we need -g, -O0 and -fPIC since we are only parsing ASTs but apart from that it seems correct.

Will those cause any issues or is it okay to keep the default behavior here?

Reply:
It should be fine.